add quantized ELU activation

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34267 add quantized ELU activation**

Summary:

Adds quantized ELU.

Test Plan:

```
python test/test_quantized.py TestQuantizedOps.test_qelu
```

still need to benchmark, saving that for after the review comments

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20370953](https://our.internmc.facebook.com/intern/diff/D20370953)

vkuzo committed:
add quantized ELU activation

Summary:

Adds quantized ELU.

Test Plan:

```
python test/test_quantized.py TestQuantizedOps.test_qelu
```

still need to benchmark, saving that for after the review comments

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] add quantized ELU activation"


Summary:

Adds quantized ELU.

Test Plan:

```
python test/test_quantized.py TestQuantizedOps.test_qelu
```

still need to benchmark, saving that for after the review comments

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

t-vi commented:
Is there a report somewhere how well this works?
Given that ELU saturates at -1 it would seem that it might run into trouble when -1 isn't represented exactly, much like ReLU needs exact representation of 0.

vkuzo commented:
> Is there a report somewhere how well this works?
> Given that ELU saturates at -1 it would seem that it might run into trouble when -1 isn't represented exactly, much like ReLU needs exact representation of 0.

yes, good point, thank you!  There is still an open TODO in this PR about the best strategy to take around scale and zero point which I should definitely improve before landing this.  Perhaps we can add optional behavior to choose output scale + zp such as both -1 and 0 are exactly represented given the data type - will look into it.

vkuzo committed:
Update on "[wip] add quantized ELU activation"


Summary:

Adds quantized ELU.

Test Plan:

```
python test/test_quantized.py TestQuantizedOps.test_qelu
```

still need to benchmark, saving that for after the review comments

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

z-a-f commented on aten/src/ATen/native/quantized/cpu/quantized_ops.h:
```diff
@@ -15,6 +15,10 @@ using qclamp_fn = void (*)(
 using qtanh_fn = void (*)(const at::Tensor& /*qx*/, at::Tensor& /*qy*/);
+using qelu_fn = void(*)(
+    const at::Tensor& /*qx*/,
+    Scalar alpha,
```
nit: remove `alpha` for consistency.

z-a-f commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -394,6 +394,79 @@ void qtanh_kernel(const Tensor& qx, Tensor& qy) {
+
+          // if (!cmp_to_zero.zero_mask()) {
+          //   // do nothing
+          // }
```
nit: cleanup

z-a-f commented on torch/nn/quantized/functional.py:
```diff
@@ -323,6 +323,36 @@ def leaky_relu(input, negative_slope=0.01, inplace=False,
     return result
 
+def elu(input, alpha=1., inplace=False, scale=None, zero_point=None):
+    r"""float(input, alpha_) -> Tensor
```
I think this line is not necessary

z-a-f reviewed (approved):
LGTM

vkuzo committed:
Update on "add quantized ELU activation"


Summary:

Adds quantized ELU.

Test Plan:

```
python test/test_quantized.py TestQuantizedOps.test_qelu
```

still need to benchmark, saving that for after the review comments

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized ELU activation"


Summary:

Adds quantized ELU.

Test Plan:

```
python test/test_quantized.py TestQuantizedOps.test_qelu
```

still need to benchmark, saving that for after the review comments

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20370953](https://our.internmc.facebook.com/intern/diff/D20370953)

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized ELU activation"


Summary:

Adds quantized ELU.

Test Plan:

```
python test/test_quantized.py TestQuantizedOps.test_qelu
```

still need to benchmark, saving that for after the review comments

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20370953](https://our.internmc.facebook.com/intern/diff/D20370953)

[ghstack-poisoned]
